// @oh-my-pi/pi-coding-agent/src/extensibility/plugins/parser.ts
/**
 * Feature bracket parser for plugin install specifiers.
 *
 * Supports syntax like:
 * - "my-plugin" -> base features (null)
 * - "npm:my-plugin" -> npm source with base features
 * - "git:github.com/user/repo" -> git source with base features
 * - "https://github.com/user/repo[search]" -> git source with features
 * - "my-plugin[search,web]" -> specific features
 * - "my-plugin[*]" -> all features
 * - "my-plugin[]" -> no optional features
 * - "@scope/plugin@1.2.3[feat]" -> scoped npm package with version and features
 */

import { parseGitUrl } from "./git-url";

export interface ParsedPluginSpec {
	/** Install target (npm package spec, git source, or supported URL) */
	target: string;
	/**
	 * Feature selection:
	 * - null: use defaults (base features on first install, preserve on reinstall)
	 * - "*": all features
	 * - string[]: specific features (empty array = no optional features)
	 */
	features: string[] | null | "*";
}

export interface ResolvedPluginInstallTarget {
	kind: "npm" | "git";
	installSpec: string;
}

function looksLikeExplicitGitSource(target: string): boolean {
	return /^git:(?!\/\/)/i.test(target) || /^(https?|ssh|git):\/\//i.test(target);
}

/**
 * Parse plugin specifier with feature bracket syntax.
 *
 * @example
 * parsePluginSpec("my-plugin") // { target: "my-plugin", features: null }
 * parsePluginSpec("my-plugin[search,web]") // { target: "my-plugin", features: ["search", "web"] }
 * parsePluginSpec("my-plugin[*]") // { target: "my-plugin", features: "*" }
 * parsePluginSpec("my-plugin[]") // { target: "my-plugin", features: [] }
 * parsePluginSpec("git:github.com/user/repo[feat]") // { target: "git:github.com/user/repo", features: ["feat"] }
 */
export function parsePluginSpec(spec: string): ParsedPluginSpec {
	// Find the last bracket pair (to handle version specifiers like @1.0.0)
	const bracketStart = spec.lastIndexOf("[");
	const bracketEnd = spec.lastIndexOf("]");

	// No brackets or malformed -> base features
	if (bracketStart === -1 || bracketEnd === -1 || bracketEnd < bracketStart) {
		return { target: spec, features: null };
	}

	const target = spec.slice(0, bracketStart);
	const featureStr = spec.slice(bracketStart + 1, bracketEnd).trim();

	// All features
	if (featureStr === "*") {
		return { target, features: "*" };
	}

	// No optional features
	if (featureStr === "") {
		return { target, features: [] };
	}

	// Specific features (comma-separated)
	const features = featureStr
		.split(",")
		.map(f => f.trim())
		.filter(Boolean);

	return { target, features };
}

/**
 * Format a parsed plugin spec back to string form.
 *
 * @example
 * formatPluginSpec({ target: "pkg", features: null }) // "pkg"
 * formatPluginSpec({ target: "pkg", features: "*" }) // "pkg[*]"
 * formatPluginSpec({ target: "pkg", features: [] }) // "pkg[]"
 * formatPluginSpec({ target: "pkg", features: ["a", "b"] }) // "pkg[a,b]"
 */
export function formatPluginSpec(spec: ParsedPluginSpec): string {
	if (spec.features === null) {
		return spec.target;
	}
	if (spec.features === "*") {
		return `${spec.target}[*]`;
	}
	if (spec.features.length === 0) {
		return `${spec.target}[]`;
	}
	return `${spec.target}[${spec.features.join(",")}]`;
}

/**
 * Normalize a plugin install target into a Bun-compatible install spec.
 *
 * Supports bare npm specs, `npm:`-prefixed npm specs, `git:` shorthand, and
 * explicit git/https URLs. Git refs are normalized to `#ref` syntax because
 * Bun accepts that form for remote installs.
 */
export function resolvePluginInstallTarget(target: string): ResolvedPluginInstallTarget {
	const trimmed = target.trim();

	if (/^npm:/i.test(trimmed)) {
		return {
			kind: "npm",
			installSpec: trimmed.slice(4).trim(),
		};
	}

	const gitSource = parseGitUrl(trimmed);
	if (gitSource) {
		return {
			kind: "git",
			installSpec: gitSource.ref ? `${gitSource.repo}#${gitSource.ref}` : gitSource.repo,
		};
	}

	if (looksLikeExplicitGitSource(trimmed)) {
		throw new Error(`Invalid git plugin source: ${target}`);
	}

	return { kind: "npm", installSpec: trimmed };
}

/**
 * Extract the installed dependency name from an npm-style package specifier.
 * Used for path lookups after npm installs.
 *
 * @example
 * extractPackageName("lodash@4.17.21") // "lodash"
 * extractPackageName("@scope/pkg@1.0.0") // "@scope/pkg"
 * extractPackageName("@scope/pkg") // "@scope/pkg"
 */
export function extractPackageName(specifier: string): string {
	// Handle scoped packages: @scope/name@version -> @scope/name
	if (specifier.startsWith("@")) {
		const match = specifier.match(/^(@[^/]+\/[^@]+)/);
		return match ? match[1] : specifier;
	}
	// Unscoped: name@version -> name
	return specifier.replace(/@[^@]+$/, "");
}
